[SYCL] Clarify sub-group size calculation in invoke_simd spec. (#6587)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_invoke_simd.asciidoc

We expect implementations of this ABI to build on Clang's `ext_vector_type`,
rather than specifying the register types directly.

Each of the following must hold true for a SIMD function and its invoke_simd
call site:

> - the value of `N` of each `simd` formal parameter corresponding to a non-uniform

Reviewer: This has to be also `simd` or `simd_mask`, correct?

Author: right, will update

> - the value of `N` of each `simd` formal parameter corresponding to a non-uniform

Author: ```suggestion
- the value of `N` of each `simd` or `simd_mask` formal parameter corresponding to a non-uniform
```

> Each of the following must hold true for a SIMD function and its invoke_simd
> call site:
> - the value of `N` of each `simd` formal parameter corresponding to a non-uniform
> SPMD actual argument (if any) must be the same
> - if the return type of the SIMD function is `simd<T, M>` or `simd_mask<T, M>`,
>   then `M` must match `N` above

Author: ```suggestion
If there are non-uniform actual arguments in `invoke_simd` call site, then each of the following must hold true for the call site and the target SIMD function:
- the value of `N` of each `simd` formal parameter corresponding to a non-uniform
SPMD actual argument (if any) must be the same
- if the return type of the SIMD function is `simd<T, M>` or `simd_mask<T, M>`,
  then `M` must match `N` above
```

Author: @Pennycook, does the above change resolve your concern?

Reviewer 2: Yes, LGTM.

Otherwise, the SIMD function cannot be invoked from an SPMD kernel and the

> implementation must issue a diagnostic in this case.

Reviewer 2: I think this is now inconsistent with the description on lines 283-294.  The text here implies that it's illegal to invoke a fully uniform function, and I prefer the description that comes later.  Do we need to duplicate it?

Author: I thought the two 'if's in the above bullets put requirements under two separate conditions, and if no conditions hold, then requirements are not imposed, and fully uniform can be called.

The lines 132-133 just try to tell what happens if the "must" requirement above does not hold, so I think this should remain in some form - ? I'll try to reformulate.

by `sub_group::get_local_id()[0]` could be passed as an integer to a
`Callable` expecting a `sycl::ext::oneapi::experimental::simd<uint32_t, N>`).

> NOTE: Implementation must be able to calculate sub-group size based on the
> `invoke_simd` SIMD call target and actual SPMD arguments. The following
> algorithm should be employed:
> - If any of the arguments is non-uniform, then the specification requires that
> the corresponding formal SIMD argument must have `simd<T, N>` or
> `simd_mask<bool, N>` type. `N` is then the sub-group size. Implementation may
> iterate through supported sub-group sizes at compile time and use
> `std::is_invocable` to see if particular sub-group size fits. There must be
> exactly one fitting sub-group size (matching `N`) or the user program is not
> well-formed.
> - Otherwise, if the return type of the SIMD target function is `simd<T, N>` or
> `simd_mask<bool, N>`, then the sub-group size is `N`.
> - Otherwise (all arguments and the return type are uniform) the SIMD target is
> fully uniform and it can be invoked with any sub-group size.

Reviewer: ```suggestion
NOTE: An implementation must compare the sub-group size to the `N` value of the SIMD
arguments of the SIMD function object to emit proper diagnostic.
The following algorithm may be employed:
- If any of the arguments is non-uniform,
the corresponding SIMD argument must have `simd<T, N>` or
`simd_mask<bool, N>` type. `N` must be identical to the the sub-group size. Implementation may
iterate through the supported sub-group sizes at compile time and check whether the
function object is invocable with a sub-group size . There must be
exactly one sub-group size (matching `N`) which allows the function object to be invoked or the user program is not
well-formed.
- Otherwise, if the return type of the SIMD target function is `simd<T, N>` or
`simd_mask<bool, N>`, then the sub-group size is `N`.
- Otherwise (all arguments and the return type are uniform) the SIMD target is
fully uniform and it can be invoked with any sub-group size.
```
